### ISSUE TITLE
Supplementary_Table_S2.py

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/loh-work-disability-experimental-notes#95]
"""Supplementary Table S2. Cluster-number and smoothing sensitivity.

Panels A-F reproduce cluster-number metrics, random-initialization stability,
the K=2/K=3 cross-tabulation, Gaussian-smoothing sensitivity of the partition
and nonlinear indicators, and nonlinear-only out-of-fold AUC sensitivity.

Prerequisites: run 00_Common_Setup.py and Figure 1.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
from scipy.ndimage import gaussian_filter1d
from scipy.stats import spearmanr
from sklearn.cluster import KMeans
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    adjusted_rand_score,
    normalized_mutual_info_score,
    roc_auc_score,
    silhouette_score,
)
from sklearn.model_selection import StratifiedKFold
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

required_globals = (
    "ANALYSIS_READY_FILE",
    "OUTPUT_DIR",
    "ANALYSIS_READY_N",
    "TRAJECTORY_COHORT_N",
    "EXPECTED_CLUSTER_COUNTS",
    "EXPECTED_ANALYSIS_SHA256",
    "RANDOM_STATE",
    "KMEANS_N_INIT",
    "read_csv_robust",
    "verify_rows",
    "verify_sha256",
    "add_analysis_row_id",
    "psa_columns_in_time_order",
)
missing_globals = [name for name in required_globals if name not in globals()]
if missing_globals:
    raise RuntimeError(
        "Run 00_Common_Setup.py before this script. Missing: "
        + ", ".join(missing_globals)
    )

FIGURE1_DIR = OUTPUT_DIR / "Figure1"
METRIC_FILE = FIGURE1_DIR / "Figure1_K2_to_K6_metrics.csv"
STABILITY_FILE = FIGURE1_DIR / "Figure1_Random_Initialization_Stability.csv"
ASSIGNMENT_FILE = FIGURE1_DIR / "Figure1_Cluster_Assignments_1434.csv"
for path in (METRIC_FILE, STABILITY_FILE, ASSIGNMENT_FILE):
    if not path.exists():
        raise FileNotFoundError(f"Missing prerequisite output: {path}")

OUTDIR = OUTPUT_DIR / "Supplementary_Table_S2"
OUTDIR.mkdir(parents=True, exist_ok=True)
SIGMA_GRID = (0.0, 0.5, 1.0, 1.5, 2.0)
REFERENCE_SIGMA_S2 = 1.0
K_GRID = (2, 3, 4, 5, 6)
S2_BOOTSTRAP = 1000

verify_sha256(ANALYSIS_READY_FILE, EXPECTED_ANALYSIS_SHA256)
source = add_analysis_row_id(read_csv_robust(ANALYSIS_READY_FILE))
verify_rows(source, ANALYSIS_READY_N, "Analysis-ready dataset")
psa_columns = psa_columns_in_time_order(source)
months = np.array([int(column.split("POM", 1)[1]) for column in psa_columns], float)
psa_numeric = source[psa_columns].apply(pd.to_numeric, errors="coerce")
positive_count = ((psa_numeric > 0) & np.isfinite(psa_numeric)).sum(axis=1)
eligible = source.loc[positive_count >= 2].copy()
verify_rows(eligible, TRAJECTORY_COHORT_N, "Eligible trajectory cohort")

def clustering_features(frame: pd.DataFrame, sigma: float) -> pd.DataFrame:
    psa = frame[psa_columns].apply(pd.to_numeric, errors="coerce")
    with np.errstate(divide="ignore", invalid="ignore"):
        log_psa = np.log(psa.replace(0, np.nan))
    adjacent_change = log_psa.diff(axis=1) / 1.0
    if sigma == 0:
        smoothed_change = adjacent_change
    else:
        smoothed_change = pd.DataFrame(
            np.vstack(
                [
                    gaussian_filter1d(row.to_numpy(float), sigma=sigma, mode="nearest")
                    for _, row in adjacent_change.iterrows()
                ]
            ),
            index=adjacent_change.index,
            columns=adjacent_change.columns,
        )
    features = pd.DataFrame(index=frame.index)
    features["max_change"] = smoothed_change.max(axis=1, skipna=True)
    features["mean_change"] = smoothed_change.mean(axis=1, skipna=True)
    features["first_positive_column_position"] = smoothed_change.apply(
        lambda row: int(np.argmax((row > 0).to_numpy(bool)))
        if np.any((row > 0).to_numpy(bool))
        else -1,
        axis=1,
    )
    return features

def imputed_matrix(features: pd.DataFrame) -> np.ndarray:
    matrix = SimpleImputer(strategy="median").fit_transform(features)
    if not np.isfinite(matrix).all():
        raise AssertionError("Non-finite clustering features remain after imputation")
    return matrix

def fit_kmeans(matrix: np.ndarray, k: int) -> tuple[KMeans, np.ndarray]:
    model = KMeans(n_clusters=k, n_init=KMEANS_N_INIT, random_state=RANDOM_STATE)
    return model, model.fit_predict(matrix)

# Panel A: K=2-6 metrics produced by the Figure 1 page.
metrics = pd.read_csv(METRIC_FILE)
panel_a = pd.DataFrame(
    {
        "K": metrics["K"].astype(int),
        "Silhouette": metrics["silhouette"],
        "Calinski-Harabasz": metrics["calinski_harabasz"],
        "Davies-Bouldin": metrics["davies_bouldin"],
        "Inertia": metrics["inertia"],
        "Reduction from preceding K, %": metrics["inertia_reduction_from_previous_pct"],
        "Sorted cluster sizes": metrics["cluster_sizes"].str.replace("|", "/", regex=False),
    }
)

# Panel B: stability across 50 random initializations.
stability = pd.read_csv(STABILITY_FILE)
panel_b = (
    stability.groupby("K", as_index=False)
    .agg(
        **{
            "Mean ARI": ("ARI_vs_seed42", "mean"),
            "Minimum ARI": ("ARI_vs_seed42", "min"),
            "Mean NMI": ("NMI_vs_seed42", "mean"),
            "Minimum NMI": ("NMI_vs_seed42", "min"),
            "Mean silhouette": ("silhouette", "mean"),
            "Minimum silhouette": ("silhouette", "min"),
            "Maximum silhouette": ("silhouette", "max"),
        }
    )
)

# Panel C: K=2 against the fixed canonical K=3 partition.
X_reference = imputed_matrix(clustering_features(eligible, REFERENCE_SIGMA_S2))
_, labels_k2 = fit_kmeans(X_reference, 2)
assignments = pd.read_csv(ASSIGNMENT_FILE)
canonical_k3 = assignments["Cluster"].to_numpy(int)
if assignments["Analysis_Row_ID"].tolist() != eligible["Analysis_Row_ID"].tolist():
    raise AssertionError("Figure 1 assignment order differs from the eligible cohort")
k2_sizes = pd.Series(labels_k2).value_counts().to_dict()
k2_name = {raw: f"K2 n={size}" for raw, size in k2_sizes.items()}
cross = pd.crosstab(pd.Series(labels_k2).map(k2_name), canonical_k3)
cross = cross.reindex(index=["K2 n=984", "K2 n=450"], columns=[1, 2, 0], fill_value=0)
panel_c = cross.rename(columns={1: "K3 C1=984", 2: "K3 C2=372", 0: "K3 C0=78"}).reset_index(names="K2 partition")
panel_c["Total"] = panel_c.iloc[:, 1:].sum(axis=1)
panel_c.loc[len(panel_c)] = ["Total", 984, 372, 78, 1434]
k2_k3_ari = float(adjusted_rand_score(labels_k2, canonical_k3))
k2_k3_nmi = float(normalized_mutual_info_score(labels_k2, canonical_k3))

# Panel D: sigma sensitivity of the K=3 partition.
_, reference_raw_labels = fit_kmeans(X_reference, 3)
sigma_cluster_rows = []
for sigma in SIGMA_GRID:
    matrix = imputed_matrix(clustering_features(eligible, sigma))
    _, labels = fit_kmeans(matrix, 3)
    counts = np.bincount(labels, minlength=3)
    sigma_cluster_rows.append(
        {
            "Gaussian sigma": sigma,
            "N": len(matrix),
            "Silhouette": silhouette_score(matrix, labels),
            "Sorted cluster sizes": "/".join(map(str, np.sort(counts))),
            "ARI versus sigma=1": adjusted_rand_score(reference_raw_labels, labels),
            "NMI versus sigma=1": normalized_mutual_info_score(reference_raw_labels, labels),
        }
    )
panel_d = pd.DataFrame(sigma_cluster_rows)

def nonlinear_indicators(values: np.ndarray, sigma: float) -> dict[str, object]:
    values = np.asarray(values, float)
    keep = np.isfinite(values) & (values > 0)
    observed_values = values[keep]
    observed_months = months[keep]
    order = np.argsort(observed_months)
    observed_values = observed_values[order]
    observed_months = observed_months[order]
    if observed_values.size < 3:
        return {"evaluable": False, "max_a": np.nan, "mean_a": np.nan, "t_star": np.nan}
    log_psa = np.log(observed_values)
    smoothed = log_psa if sigma == 0 else gaussian_filter1d(log_psa, sigma=sigma, mode="nearest")
    a_t = -np.gradient(smoothed, observed_months)
    da_dt = np.gradient(a_t, observed_months)
    if not (np.isfinite(a_t).all() and np.isfinite(da_dt).all()):
        return {"evaluable": False, "max_a": np.nan, "mean_a": np.nan, "t_star": np.nan}
    peak_index = int(np.argmax(da_dt))
    return {
        "evaluable": True,
        "max_a": float(np.max(a_t)),
        "mean_a": float(np.mean(a_t)),
        "t_star": float(observed_months[peak_index]),
    }

descriptor_frames = {}
for sigma in SIGMA_GRID:
    records = []
    for _, row in eligible.iterrows():
        record = nonlinear_indicators(
            pd.to_numeric(row[psa_columns], errors="coerce").to_numpy(float), sigma
        )
        record["Analysis_Row_ID"] = int(row["Analysis_Row_ID"])
        records.append(record)
    descriptor_frames[sigma] = pd.DataFrame(records)

def safe_spearman(first: pd.Series, second: pd.Series) -> float:
    pair = pd.DataFrame({"first": first, "second": second}).dropna()
    return float(spearmanr(pair["first"], pair["second"]).statistic)

# Panel E: nonlinear-indicator sensitivity versus sigma=1.
reference_descriptors = descriptor_frames[REFERENCE_SIGMA_S2]
descriptor_rows = []
for sigma in SIGMA_GRID:
    current = descriptor_frames[sigma].merge(
        reference_descriptors,
        on="Analysis_Row_ID",
        suffixes=("", "_reference"),
        validate="one_to_one",
    )
    common = current["evaluable"].astype(bool) & current["evaluable_reference"].astype(bool)
    paired = current.loc[common]
    shift = np.abs(paired["t_star"] - paired["t_star_reference"])
    descriptor_rows.append(
        {
            "Gaussian sigma": sigma,
            "Evaluable, n": int(current["evaluable"].sum()),
            "rho max a(t)": safe_spearman(paired["max_a"], paired["max_a_reference"]),
            "rho mean a(t)": safe_spearman(paired["mean_a"], paired["mean_a_reference"]),
            "rho t*": safe_spearman(paired["t_star"], paired["t_star_reference"]),
            "Exact t* agreement, %": 100 * np.mean(shift == 0),
            "Within 3 months, %": 100 * np.mean(shift <= 3),
            "Median absolute shift, months": float(np.median(shift)),
        }
    )
panel_e = pd.DataFrame(descriptor_rows)

# Panel F: nonlinear-only out-of-fold AUC against the fixed sigma=1 C1 target.
target = assignments.copy()
target["target_C1"] = (target["Cluster"] == 1).astype(int)
y = target["target_C1"].to_numpy(int)
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
splits = list(cv.split(np.zeros(len(target)), y))

def nonlinear_model() -> Pipeline:
    return Pipeline(
        [
            ("imputer", SimpleImputer(strategy="median")),
            ("scaler", StandardScaler()),
            ("classifier", LogisticRegression(max_iter=5000, solver="lbfgs", random_state=RANDOM_STATE)),
        ]
    )

probabilities = {}
auc_rows = []
features = ["t_star", "max_a", "mean_a"]
for sigma in SIGMA_GRID:
    data = target.merge(
        descriptor_frames[sigma][["Analysis_Row_ID", *features]],
        on="Analysis_Row_ID",
        validate="one_to_one",
    )
    probability = np.full(len(data), np.nan)
    for train_index, test_index in splits:
        model = nonlinear_model()
        model.fit(data.iloc[train_index][features], y[train_index])
        probability[test_index] = model.predict_proba(data.iloc[test_index][features])[:, 1]
    probabilities[sigma] = probability
    auc_rows.append({"Gaussian sigma": sigma, "N": len(data), "ROC AUC": roc_auc_score(y, probability)})

panel_f = pd.DataFrame(auc_rows)
reference_probability = probabilities[REFERENCE_SIGMA_S2]
for index, row in panel_f.iterrows():
    sigma = float(row["Gaussian sigma"])
    rng = np.random.default_rng(RANDOM_STATE + int(round(sigma * 100)))
    differences = []
    for _ in range(S2_BOOTSTRAP):
        sample = rng.integers(0, len(y), len(y))
        sampled_y = y[sample]
        if np.unique(sampled_y).size < 2:
            continue
        differences.append(
            roc_auc_score(sampled_y, probabilities[sigma][sample])
            - roc_auc_score(sampled_y, reference_probability[sample])
        )
    differences = np.asarray(differences, float)
    panel_f.loc[index, "Bootstrap mean difference versus sigma=1"] = np.mean(differences)
    panel_f.loc[index, "95% CI lower"] = np.percentile(differences, 2.5)
    panel_f.loc[index, "95% CI upper"] = np.percentile(differences, 97.5)

# Manuscript-value completion gates (rounded as displayed in Supplementary Table S2).
expected_a = [
    (2, .936, 25902.4, .157, 745.6, "450/984"),
    (3, .914, 42022.9, .370, 238.3, "78/372/984"),
    (4, .896, 42313.8, .474, 158.5, "41/82/327/984"),
    (5, .878, 42901.7, .520, 117.5, "30/41/82/327/954"),
    (6, .865, 44353.9, .506, 91.1, "23/30/45/55/327/954"),
]
observed_a = [
    (
        int(row["K"]),
        round(float(row["Silhouette"]), 3),
        round(float(row["Calinski-Harabasz"]), 1),
        round(float(row["Davies-Bouldin"]), 3),
        round(float(row["Inertia"]), 1),
        row["Sorted cluster sizes"],
    )
    for _, row in panel_a.iterrows()
]
if observed_a != expected_a:
    raise AssertionError(f"Panel A mismatch: {observed_a}")
if (round(k2_k3_ari, 3), round(k2_k3_nmi, 3)) != (0.943, 0.896):
    raise AssertionError(f"Panel C mismatch: ARI={k2_k3_ari}, NMI={k2_k3_nmi}")
if panel_d["Sorted cluster sizes"].tolist() != ["185/225/1024", "196/607/631", "78/372/984", "10/177/1247", "9/17/1408"]:
    raise AssertionError("Panel D cluster sizes do not match the manuscript")
if [round(value, 3) for value in panel_f["ROC AUC"]] != [.777, .769, .735, .720, .716]:
    raise AssertionError("Panel F AUC values do not match the manuscript")

panels = {"A": panel_a, "B": panel_b, "C": panel_c, "D": panel_d, "E": panel_e, "F": panel_f}
for label, frame in panels.items():
    frame.to_csv(
        OUTDIR / f"Supplementary_Table_S2{label}.csv",
        index=False,
        encoding="utf-8-sig",
    )

markdown_parts = [
    "# Supplementary Table S2. Cluster-number and Gaussian-smoothing sensitivity analyses of postoperative PSA trajectory partitions",
]
panel_titles = {
    "A": "Cluster-number metrics",
    "B": "Random-initialization stability",
    "C": f"K=2 versus K=3 cross-tabulation (ARI={k2_k3_ari:.3f}; NMI={k2_k3_nmi:.3f})",
    "D": "Gaussian-smoothing sensitivity of cluster membership",
    "E": "Gaussian-smoothing sensitivity of nonlinear indicators",
    "F": "Nonlinear-only out-of-fold AUC sensitivity",
}
for label, frame in panels.items():
    markdown_parts.extend([f"## Panel {label}. {panel_titles[label]}", frame.to_markdown(index=False)])
(OUTDIR / "Supplementary_Table_S2.md").write_text("\n\n".join(markdown_parts) + "\n", encoding="utf-8")

print("SUPPLEMENTARY TABLE S2 COMPLETION GATE PASSED")
for label, frame in panels.items():
    print(f"\nPanel {label}\n", frame.to_string(index=False))
